src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -0,0 +1,432 @@
+        } else {
+            LOGGER.log(WARNING, "cannot cancel dead #" + build);
+        }
+    }
+
+    private static void cleanUp(Job<?,?> job, String jobName) {
+        Map<Integer, Milestone> milestonesInJob = getMilestonesByOrdinalByJob().get(jobName);
+        assert milestonesInJob != null;
+        Iterator<Entry<Integer, Milestone>> it = milestonesInJob.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<Integer, Milestone> entry = it.next();
+            Set<Integer> inSight = entry.getValue().inSight;
+            Iterator<Integer> it2 = inSight.iterator();
+            while (it2.hasNext()) {
+                Integer number = it2.next();
+                if (job.getBuildByNumber(number) == null) {
+                    // Deleted at some point but did not properly clean up from exit(…).
+                    LOGGER.log(WARNING, "Cleaning up apparently deleted {0}#{1}", new Object[] {jobName, number});
+                    it2.remove();
+                }
+            }
+        }
+        // TODO: remove milestones that no longer exist (because the step or the job was deleted)
+    }
+
+    @SuppressWarnings("unchecked")
+    private static synchronized void load() {
+        if (getMilestonesByOrdinalByJob() == null) {
+            MilestoneStep.DescriptorImpl.milestonesByOrdinalByJob = new TreeMap<String, Map<Integer, Milestone>>();
+            try {
+                XmlFile configFile = getConfigFile();
+                if (configFile.exists()) {
+                    MilestoneStep.DescriptorImpl.milestonesByOrdinalByJob = (Map<String, Map<Integer, Milestone>>) configFile.read();
+                }
+            } catch (IOException x) {
+                LOGGER.log(WARNING, null, x);
+            }
+            LOGGER.log(Level.FINE, "load: {0}", getMilestonesByOrdinalByJob());
+        }
+    }
+
+    private static synchronized void save() {
+        try {
+            getConfigFile().write(getMilestonesByOrdinalByJob());
+        } catch (IOException x) {
+            LOGGER.log(WARNING, null, x);
+        }
+        LOGGER.log(Level.FINE, "save: {0}", getMilestonesByOrdinalByJob());
+    }
+
+    private static XmlFile getConfigFile() throws IOException {
+        Jenkins j = Jenkins.getInstance();
+        if (j == null) {
+            throw new IOException("Jenkins is not running"); // do not use Jenkins.getActiveInstance() as that is an ISE
+        }
+        return new XmlFile(new File(j.getRootDir(), MilestoneStep.class.getName() + ".xml"));
+    }
+
+    @Extension
+    public static final class Listener extends RunListener<Run<?,?>> {
+        @Override public void onCompleted(Run<?,?> r, TaskListener listener) {
+            if (!(r instanceof FlowExecutionOwner.Executable) || ((FlowExecutionOwner.Executable) r).asFlowExecutionOwner() == null) {
+                return;
+            }
+            exit(r);
+        }
+    }
+
+    static final class Milestone {
```

Comment:
:ant: If you are going to make a dedicated plugin, you may as well make this a top-level type.